File: go/2024/19/19.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("in.txt")
	split := strings.Split(strings.TrimSpace(string(input)), "\n\n")
	patterns := strings.Split(split[0], ", ")
	targets := strings.Split(split[1], "\n")
	fmt.Println(patterns)
	fmt.Println(targets)

	p1 := 0
	p2 := 0
	cache := map[string]int{}
	var ways func(string, []string) int

	ways = func(target string, patterns []string) int {
		if n, ok := cache[target]; ok {
			return n
		}
		n := 0
		if len(target) == 0 {
			n = 1
		}
		for _, p := range patterns {
			if strings.HasPrefix(target, p) {
				n += ways(target[len(p):], patterns)
			}
		}
		cache[target] = n
		return n
	}

	for _, t := range targets {
		n := ways(t, patterns)
		if n > 0 {
			p1++
		}
		p2 += n
	}
	fmt.Println(p1, p2)
}

File: python/2024/19.py
a, b = open("in.txt").read().split("\n\n")
patterns = a.split(", ")
targets = b.splitlines()


## Method 1
from functools import cache


@cache
def is_possible(target, op):
    return not target or op(
        is_possible(target[len(p) :], op) for p in patterns if target.startswith(p)
    )


for op in any, sum:
    print(sum(is_possible(target, op) for target in targets))


## Method 2
cache = {}


def ways(patterns, target):
    if target in cache:
        return cache[target]
    ans = 0
    if not target:
        ans = 1
    for p in patterns:
        if target.startswith(p):
            ans += ways(patterns, target[len(p) :])
    cache[target] = ans
    return ans


p1 = p2 = 0

for target in targets:
    target_ways = ways(patterns, target)
    if target_ways > 0:
        p1 += 1
    p2 += target_ways

print(p1, p2)
